controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,208 @@
+
+	for _, deviceSet := range storageCluster.Spec.StorageDeviceSets {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		if (deviceSet.DeviceClass == "" && deviceClass == "ssd") || deviceSet.DeviceClass == deviceClass {
+			startOsdSize = deviceSet.DataPVCTemplate.Spec.Resources.Requests["storage"]
+			startOsdCount = deviceSet.Count
+			deviceSetCount++
+		}
+	}
+
+	startStorageCapacity := *resource.NewQuantity(int64(deviceSetCount*startOsdCount)*startOsdSize.Value(), startOsdSize.Format)
+
+	if startOsdSize.Cmp(storageAutoScaler.Spec.MaxOsdSize) < 0 {
+		// vertical scaling
+		// double the osd size
+		expectedOsdSize = *resource.NewQuantity(startOsdSize.Value()*2, startOsdSize.Format)
+	} else {
+		// horizontal scaling
+		expectedOsdCount = startOsdCount + 1
+	}
+
+	expectedStorageCapacity := *resource.NewQuantity(int64(deviceSetCount*expectedOsdCount)*expectedOsdSize.Value(), startOsdSize.Format)
+
+	if expectedStorageCapacity.Cmp(storageAutoScaler.Spec.StorageCapacityLimit) > 0 {
+		r.Log.Info("storage capacity limit reached")
+		// update the status
+		bool := true
+		storageAutoScaler.Status.StorageCapacityLimitReached = &bool
+		err := r.updateStatus(ctx, storageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// update the status
+	storageAutoScaler.Status.Phase = "in-progress"
+	storageAutoScaler.Status.LastExpansion = &ocsv1.LastExpansionStatus{
+		StartOsdCount:           uint16(startOsdCount),
+		ExpectedOsdCount:        uint16(expectedOsdCount),
+		StartOsdSize:            startOsdSize,
+		ExpectedOsdSize:         expectedOsdSize,
+		StartStorageCapacity:    startStorageCapacity,
+		ExpectedStorageCapacity: expectedStorageCapacity,
+		StartTime:               metav1.Now(),
+	}
+	err = r.updateStatus(ctx, storageAutoScaler)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// patch the storagecluster with the expected osd size and count
+	for _, deviceSet := range storageCluster.Spec.StorageDeviceSets {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		if (deviceSet.DeviceClass == "" && deviceClass == "ssd") || deviceSet.DeviceClass == deviceClass {
+			deviceSet.DataPVCTemplate.Spec.Resources.Requests["storage"] = expectedOsdSize
+			deviceSet.Count = expectedOsdCount
+		}
+	}
+
+	err = r.Update(ctx, storageCluster)
+	if err != nil {
+		r.Log.Error(err, "failed to update storage cluster")
+		return reconcile.Result{}, err
+	}
+
+	return r.verifyScaling()
+}
+
+func (r *StorageAutoscalerReconciler) updateStatus(ctx context.Context, storageAutoScaler *ocsv1.StorageAutoScaler) error {
+	err := r.Status().Update(ctx, storageAutoScaler)
+	if err != nil {
+		r.Log.Error(err, "failed to update storage autoscaler status")
+		return err
+	}
+	r.Log.Info("storage autoscaler status updated")
+	return nil
+}
+
+func (r *StorageAutoscalerReconciler) verifyScaling() (reconcile.Result, error) {
+	// verify the scaling
```

Comment:
Should we query the prometheus to verify it? 
cc @travisn

Reply:
Yes, I see querying prometheus as the correct approach to verifying that the scaling is completed. If we used a different approach, then we might be at risk of verifying the wrong thing. By using the same approach to both trigger and verify the expansion, it will be most reliable.